sdc/native/hashmap.hpp

```diff
@@ -0,0 +1,862 @@
+        // moving to one impl later if there's no performance penalty
+        auto p_hash_map = new numeric_hashmap<key_type, val_type>;
+        (*meminfo) = nrt->manage_memory((void*)p_hash_map, delete_numeric_hashmap<key_type, val_type>);
+    }
+
+    return;
+}
+
+
+template<typename key_type, typename val_type>
+uint64_t hashmap_size(void* p_hash_map)
+{
+    auto p_hash_map_spec = reinterpet_hashmap_ptr<key_type, val_type>(p_hash_map);
+    return p_hash_map_spec->size();
+}
+
+
+template<typename key_type, typename val_type>
+void hashmap_set(void* p_hash_map, key_type key, val_type val)
+{
+    auto p_hash_map_spec = reinterpet_hashmap_ptr<key_type, val_type>(p_hash_map);
+    p_hash_map_spec->set(key, val);
+}
+
+
+template<typename key_type, typename val_type>
+int8_t hashmap_lookup(void* p_hash_map,
+                      key_type key,
+                      val_type* res)
+{
+    auto p_hash_map_spec = reinterpet_hashmap_ptr<key_type, val_type>(p_hash_map);
+    return p_hash_map_spec->lookup(key, res);
+}
+
+
+template<typename key_type, typename val_type>
+void hashmap_clear(void* p_hash_map)
+{
+    auto p_hash_map_spec = reinterpet_hashmap_ptr<key_type, val_type>(p_hash_map);
+    p_hash_map_spec->clear();
+}
+
+
+template<typename key_type, typename val_type>
+int8_t hashmap_unsafe_extract(void* p_hash_map, key_type key, val_type* res)
+{
+    auto p_hash_map_spec = reinterpet_hashmap_ptr<key_type, val_type>(p_hash_map);
+    return p_hash_map_spec->pop(key, res);
+}
+
+
+template<typename key_type, typename val_type>
+void hashmap_numeric_from_arrays(NRT_MemInfo** meminfo, void* nrt_table, key_type* keys, val_type* values, uint64_t size)
+{
+    auto nrt = (NRT_api_functions*)nrt_table;
+    auto p_hash_map = new numeric_hashmap<key_type, val_type>;
+    (*meminfo) = nrt->manage_memory((void*)p_hash_map, delete_numeric_hashmap<key_type, val_type>);
+
+    // FIXME: apply arena to make this react on changing NUMBA_NUM_THREADS
+    tbb::parallel_for(tbb::blocked_range<size_t>(0, size),
+                 [=](const tbb::blocked_range<size_t>& r) {
+                     for(size_t i=r.begin(); i!=r.end(); ++i) {
+                         auto kv_pair = std::pair<const key_type, val_type>(keys[i], values[i]);
+                         p_hash_map->map.insert(
+                             std::move(kv_pair)
+                         );
+                     }
+                 }
+    );
+}
+
+
+template<typename key_type, typename val_type>
+void hashmap_update(void* p_self_hash_map, void* p_arg_hash_map)
+{
+    auto p_self_hash_map_spec = reinterpet_hashmap_ptr<key_type, val_type>(p_self_hash_map);
+    auto p_arg_hash_map_spec = reinterpet_hashmap_ptr<key_type, val_type>(p_arg_hash_map);
+    p_self_hash_map_spec->update(*p_arg_hash_map_spec);
+    return;
+}
+
+
+template<typename key_type, typename val_type>
+void hashmap_dump(void* p_hash_map)
```

Comment:
Can we make  `hashmap_dump` and corresponding `#include <iostream>` optional and wrap in some ifdefs? I would really prefer release libraries without iostream dependency?

Reply:
Done.